ambry-clustermap/src/main/java/com/github/ambry/clustermap/PropertyStoreToDataNodeConfigAdapter.java

```diff
@@ -0,0 +1,276 @@
+          } else {
+            LOGGER.info("DataNodeConfig at path {} not found", changedPath);
+          }
+        }
+      } catch (Exception e) {
+        LOGGER.error("Exception during DataNodeConfig change", e);
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    /**
+     * If initialization has not yet been completed succesfully, attempt an initialization. This should be called from
+     * the event executor thread.
+     * @return {@code true} if a successful full initialization (whether successful or not) was attempted in this call.
+     */
+    private boolean initializeIfNeeded() {
+      lock.lock();
+      try {
+        if (initialized) {
+          return false;
+        }
+        List<ZNRecord> records = propertyStore.getChildren(CONFIG_PATH, null, AccessOption.PERSISTENT);
+        listener.onDataNodeConfigChange(lazyIterable(records));
+        initialized = true;
+      } catch (Exception e) {
+        LOGGER.error("Exception during DataNodeConfig initialization", e);
+      } finally {
+        lock.unlock();
+      }
+      return true;
+    }
+  }
+
+  /**
+   * Convert between {@link DataNodeConfig}s and {@link ZNRecord}s.
+   */
+  static class Converter {
+    private static final int VERSION_0 = 0;
+    private static final String MOUNT_PREFIX = "mount-";
+    private static final String HOSTNAME_FIELD = "hostname";
+    private static final String PORT_FIELD = "port";
+    private final String defaultPartitionClass;
+    private final String dcName;
+
+    /**
+     * @param defaultPartitionClass the default partition class for these nodes.
+     * @param dcName the datacenter name for these nodes.
+     */
+    Converter(String defaultPartitionClass, String dcName) {
+      this.defaultPartitionClass = defaultPartitionClass;
+      this.dcName = dcName;
+    }
+
+    /**
+     * @param record the {@link ZNRecord} to convert to a {@link DataNodeConfig} object.
+     * @return the {@link DataNodeConfig}, or {@code null} if the {@link ZNRecord} provided has an unsupported schema
+     *         version.
+     */
+    DataNodeConfig convert(ZNRecord record) {
+      int schemaVersion = getSchemaVersion(record);
+      if (schemaVersion != VERSION_0) {
+        LOGGER.warn("Unknown InstanceConfig schema version {} in {}. Ignoring.", schemaVersion, record);
+        return null;
+      }
+
+      DataNodeConfig dataNodeConfig = new DataNodeConfig(record.getId(), record.getSimpleField(HOSTNAME_FIELD),
+          record.getIntField(PORT_FIELD, DataNodeId.UNKNOWN_PORT), dcName, getSslPortStr(record),
+          getHttp2PortStr(record), getRackId(record), DEFAULT_XID);
+      dataNodeConfig.getSealedReplicas().addAll(getSealedReplicas(record));
+      dataNodeConfig.getStoppedReplicas().addAll(getStoppedReplicas(record));
+      dataNodeConfig.getDisabledReplicas().addAll(getDisabledReplicas(record));
+      record.getMapFields().forEach((key, diskProps) -> {
+        if (key.startsWith(MOUNT_PREFIX)) {
```

Comment:
Looks like we will change the format of mount path in PropertyStore record?  Unlike InstanceConfig (which is a pure mount path), here we add a prefix `mount-` ahead of each mount path. I just wonder why not use the same in `InstanceConfigToDataNodeConfigAdapter` to determine if this is a valid mount path.

Reply:
The way that we had it would work correctly too, but I felt that was a little bit of a hacky approach since it checked the content inside of the map field. My thought around adding a prefix was that it would give us more flexibility to add other types of map fields in the future. 

Since there are a few other changes to the schema, like the removal of datacenter name and xid, I thought this was a good time to make changes. However, if you prefer the shorter name, I could go back to that too.

Reply:
Make sense, let's keep the prefix there.